[Console] more secondary output choices (try 2) (IDFGH-8688)

chipweinberger opened the pull request:
🥳🥳🥳🥳

original PR: https://github.com/espressif/esp-idf/pull/9979

Okay. @igrr Made some big changes to the diff.

I combined the primary & secondary UART, like you suggested. It's a good change.

For clarity, I added internal KConfigs for when either primary or secondary is enabled, i.e. `ESP_CONSOLE_IS_USB_CDC_ENABLED`

I've tested:

- building S2, all 6-ish permutations
- building S3, all 14-ish permutations (CDC does not support dual core yet)
- Running the Console Example for S2, UART, UART + CDC, (I don't have a USB port to test CDC alone)
- Running the Console Example for S3, UART, Serial JTAG, UART + Serial JTAG, Serial JTAG + UART

It should be good to merge.

igrr commented on components/bootloader_support/src/bootloader_console.c:
```diff
@@ -109,12 +118,34 @@ void bootloader_console_init(void)
+#if CONFIG_ESP_CONSOLE_IS_USB_SERIAL_JTAG_ENABLED
+    bootloader_console_init_usb_serial_jtag();
+#endif
+}
```
Looks like here we may end up calling esp_rom_install_channel_putc (for redirecting ets_printf/esp_rom_printf) multiple times, and it will get redirected to the most recent destination. Is that the expected behavior?

chipweinberger replied:
Which 2 functions both call `esp_rom_install_channel_putc()`? 

To be honest, I'm not familiar with what "redirecting ets_printf" means. 

Regardless, I only see that`bootloader_console_init_none()` and `bootloader_console_init_usb_cdc() `call `esp_rom_install_channel_putc()`, but `NONE` is only true when the others are NOT enabled, so it is not a problem

igrr reviewed (commented):
One comment, and please check the pre-commit as well ([docs](https://docs.espressif.com/projects/esp-idf/en/latest/esp32/contribute/install-pre-commit-hook.html)), it is still failing.

chipweinberger commented:
not having much luck with the precommit. lots of errors.

<details>
  <summary>Click me</summary>
  
```
# Please combine this with the original one
#
# Notes:
# - please keep in mind to avoid duplicated folders as yaml keys
# - please use parentheses to group conditions, the "or" and "and" operators could only accept two operands
components/esp_system/test_apps/rtc_8md256:
  enable:
    - if: IDF_TARGET in ["esp32", "esp32c2", "esp32c3", "esp32s2", "esp32s3"]
      temporary: true
      reason: <why only enable build jobs for these targets>
check_enable_build failed for app: (cmake) App components/esp_system/test_apps/rtc_8md256, target esp32, sdkconfig (default), build in /Users/charlesweinberger/Desktop/esp-idf-v5/esp-idf/components/esp_system/test_apps/rtc_8md256/build
--------------------------------------------------------------------------------
Modified README file: components/esp_system/test_apps/rtc_8md256/README.md
--------------------------------------------------------------------------------
components/esp_system/test_apps/rtc_power_modes:
- enable build targets according to the manifest file: ['esp32', 'esp32s2', 'esp32c3', 'esp32s3', 'esp32c2', 'esp32c6']
- enable build targets according to the old Supported Targets table under readme "components/esp_system/test_apps/rtc_power_modes/README.md": ['esp32s3']

If you want to disable some targets, please use the following snippet:

# Please combine this with the original one
#
# Notes:
# - please keep in mind to avoid duplicated folders as yaml keys
# - please use parentheses to group conditions, the "or" and "and" operators could only accept two operands
components/esp_system/test_apps/rtc_power_modes:
  enable:
    - if: IDF_TARGET in ["esp32s3"]
      temporary: true
      reason: <why only enable build jobs for these targets>
check_enable_build failed for app: (cmake) App components/esp_system/test_apps/rtc_power_modes, target esp32, sdkconfig (default), build in /Users/charlesweinberger/Desktop/esp-idf-v5/esp-idf/components/esp_system/test_apps/rtc_power_modes/build
--------------------------------------------------------------------------------
Modified README file: components/esp_system/test_apps/rtc_power_modes/README.md
--------------------------------------------------------------------------------

sort .build-test-rules.yml files.............................................................................(no files to check)Skipped
File Contents Sorter.........................................................................................(no files to check)Skipped
Check copyright notices..........................................................................................................Passed
charlesweinberger@Chips-Macbook-Pro-2 esp-idf % git commit -m "[Console] more secondary output choices"
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /Users/charlesweinberger/.cache/pre-commit/patch1667903318-74132.
Trim Trailing Whitespace.........................................................................................................Passed
Fix End of Files.................................................................................................................Passed
Check that executables have shebangs.........................................................................(no files to check)Skipped
Mixed line ending................................................................................................................Passed
Fix double quoted strings....................................................................................(no files to check)Skipped
flake8.......................................................................................................(no files to check)Skipped
isort (python)...............................................................................................(no files to check)Skipped
Check File Permissions.......................................................................................(no files to check)Skipped
Validate executable-list.txt.....................................................................................................Failed
- hook id: check-executable-list
- exit code: 1

/Volumes/User/MBP-Google-Drive/jamcorder/firmware/esp-idf/tools/ci/executable-list.txt:1 .gitlab/ci/dependencies/generate_rules.py not exists. Please remove it manually

Validate Kconfig files...........................................................................................................Passed
Check if any Kconfig Options Deprecated......................................................................(no files to check)Skipped
Check CMake Files Format.........................................................................................................Passed
Validate Codeowner File..........................................................................................................Failed
- hook id: check-codeowners
- exit code: 1

Errors found!
/Volumes/User/MBP-Google-Drive/jamcorder/firmware/esp-idf/.gitlab/CODEOWNERS:56: no files matched by pattern /CONTRIBUTING.md
/Volumes/User/MBP-Google-Drive/jamcorder/firmware/esp-idf/.gitlab/CODEOWNERS:61: no files matched by pattern /conftest.py
/Volumes/User/MBP-Google-Drive/jamcorder/firmware/esp-idf/.gitlab/CODEOWNERS:64: no files matched by pattern /pytest.ini
/Volumes/User/MBP-Google-Drive/jamcorder/firmware/esp-idf/.gitlab/CODEOWNERS:73: no files matched by pattern /components/bootloader_support/bootloader_flash/
/Volumes/User/MBP-Google-Drive/jamcorder/firmware/esp-idf/.gitlab/CODEOWNERS:80: no files matched by pattern /components/esp_adc/
/Volumes/User/MBP-Google-Drive/jamcorder/firmware/esp-idf/.gitlab/CODEOWNERS:81: no files matched by pattern /components/esp_app_format/
/Volumes/User/MBP-Google-Drive/jamcorder/firmware/esp-idf/.gitlab/CODEOWNERS:97: no files matched by pattern /components/esp_psram/
/Volumes/User/MBP-Google-Drive/jamcorder/firmware/esp-idf/.gitlab/CODEOWNERS:107: no files matched by pattern /components/fatfs/**/*.py
/Volumes/User/MBP-Google-Drive/jamcorder/firmware/esp-idf/.gitlab/CODEOWNERS:111: no files matched by pattern /components/http_parser/
/Volumes/User/MBP-Google-Drive/jamcorder/firmware/esp-idf/.gitlab/CODEOWNERS:153: no files matched by pattern /docs/**/migration-guides/
/Volumes/User/MBP-Google-Drive/jamcorder/firmware/esp-idf/.gitlab/CODEOWNERS:176: no files matched by pattern /examples/zigbee/
/Volumes/User/MBP-Google-Drive/jamcorder/firmware/esp-idf/.gitlab/CODEOWNERS:199: no files matched by pattern /tools/test_build_system/
/Volumes/User/MBP-Google-Drive/jamcorder/firmware/esp-idf/.gitlab/CODEOWNERS:208: no files matched by pattern /tools/ci/check_copyright_config.yaml
/Volumes/User/MBP-Google-Drive/jamcorder/firmware/esp-idf/.gitlab/CODEOWNERS:209: no files matched by pattern /tools/ci/check_copyright_ignore.txt

Check rules.yml all rules have at lease one job applied, all rules needed exist..............................(no files to check)Skipped
Check rules are generated (based on .gitlab/ci/dependencies/dependencies.yml)................................(no files to check)Skipped
Check type annotations in python files.......................................................................(no files to check)Skipped
Check requirement files......................................................................................(no files to check)Skipped
Check tools dir files patterns...............................................................................(no files to check)Skipped
Check soc caps kconfig files are generated (based on components/soc/IDF_TARGET/include/soc/soc_caps.h).......(no files to check)Skipped
Check if all apps readme files match given .build-test-rules.yml files. Modify the supported target tables.......................Failed
- hook id: check-all-apps-readmes
- exit code: 1
- files were modified by this hook

components/console/test_apps/console:
- enable build targets according to the manifest file: ['esp32', 'esp32s2', 'esp32c3', 'esp32s3', 'esp32c2', 'esp32c6']
- enable build targets according to the old Supported Targets table under readme "components/console/test_apps/console/README.md": ['esp32', 'esp32c2', 'esp32c3', 'esp32c6', 'esp32s2', 'esp32s3']

If you want to disable some targets, please use the following snippet:

# Please combine this with the original one
#
# Notes:
# - please keep in mind to avoid duplicated folders as yaml keys
# - please use parentheses to group conditions, the "or" and "and" operators could only accept two operands
components/console/test_apps/console:
  enable:
    - if: IDF_TARGET in ["esp32", "esp32c2", "esp32c3", "esp32c6", "esp32s2", "esp32s3"]
      temporary: true
      reason: <why only enable build jobs for these targets>
check_enable_build failed for app: (cmake) App components/console/test_apps/console, target esp32, sdkconfig (default), build in /Users/charlesweinberger/Desktop/esp-idf-v5/esp-idf/components/console/test_apps/console/build
--------------------------------------------------------------------------------
Modified README file: components/console/test_apps/console/README.md
--------------------------------------------------------------------------------
components/esp_system/test_apps/rtc_8md256:
- enable build targets according to the manifest file: ['esp32', 'esp32s2', 'esp32c3', 'esp32s3', 'esp32c2', 'esp32c6']
- enable build targets according to the old Supported Targets table under readme "components/esp_system/test_apps/rtc_8md256/README.md": ['esp32', 'esp32s2', 'esp32c3', 'esp32s3', 'esp32c2', 'esp32c6']

If you want to disable some targets, please use the following snippet:

# Please combine this with the original one
#
# Notes:
# - please keep in mind to avoid duplicated folders as yaml keys
# - please use parentheses to group conditions, the "or" and "and" operators could only accept two operands
components/esp_system/test_apps/rtc_8md256:
  enable:
    - if: IDF_TARGET in ["esp32", "esp32c2", "esp32c3", "esp32c6", "esp32s2", "esp32s3"]
      temporary: true
      reason: <why only enable build jobs for these targets>
check_enable_build failed for app: (cmake) App components/esp_system/test_apps/rtc_8md256, target esp32, sdkconfig (default), build in /Users/charlesweinberger/Desktop/esp-idf-v5/esp-idf/components/esp_system/test_apps/rtc_8md256/build
--------------------------------------------------------------------------------
components/esp_system/test_apps/rtc_power_modes:
- enable build targets according to the manifest file: ['esp32', 'esp32s2', 'esp32c3', 'esp32s3', 'esp32c2', 'esp32c6']
- enable build targets according to the old Supported Targets table under readme "components/esp_system/test_apps/rtc_power_modes/README.md": ['esp32', 'esp32s2', 'esp32c3', 'esp32s3', 'esp32c2', 'esp32c6']

If you want to disable some targets, please use the following snippet:

# Please combine this with the original one
#
# Notes:
# - please keep in mind to avoid duplicated folders as yaml keys
# - please use parentheses to group conditions, the "or" and "and" operators could only accept two operands
components/esp_system/test_apps/rtc_power_modes:
  enable:
    - if: IDF_TARGET in ["esp32", "esp32c2", "esp32c3", "esp32c6", "esp32s2", "esp32s3"]
      temporary: true
      reason: <why only enable build jobs for these targets>
check_enable_build failed for app: (cmake) App components/esp_system/test_apps/rtc_power_modes, target esp32, sdkconfig (default), build in /Users/charlesweinberger/Desktop/esp-idf-v5/esp-idf/components/esp_system/test_apps/rtc_power_modes/build
--------------------------------------------------------------------------------
components/esp_system/test_apps/rtc_8md256:
- enable build targets according to the manifest file: ['esp32', 'esp32s2', 'esp32c3', 'esp32s3', 'esp32c2', 'esp32c6']
- enable build targets according to the old Supported Targets table under readme "components/esp_system/test_apps/rtc_8md256/README.md": ['esp32', 'esp32s2', 'esp32c3', 'esp32s3', 'esp32c2', 'esp32c6']

If you want to disable some targets, please use the following snippet:

# Please combine this with the original one
#
# Notes:
# - please keep in mind to avoid duplicated folders as yaml keys
# - please use parentheses to group conditions, the "or" and "and" operators could only accept two operands
components/esp_system/test_apps/rtc_8md256:
  enable:
    - if: IDF_TARGET in ["esp32", "esp32c2", "esp32c3", "esp32c6", "esp32s2", "esp32s3"]
      temporary: true
      reason: <why only enable build jobs for these targets>
check_enable_build failed for app: (cmake) App components/esp_system/test_apps/rtc_8md256, target esp32, sdkconfig (default), build in /Users/charlesweinberger/Desktop/esp-idf-v5/esp-idf/components/esp_system/test_apps/rtc_8md256/build
--------------------------------------------------------------------------------
components/esp_system/test_apps/rtc_power_modes:
- enable build targets according to the manifest file: ['esp32', 'esp32s2', 'esp32c3', 'esp32s3', 'esp32c2', 'esp32c6']
- enable build targets according to the old Supported Targets table under readme "components/esp_system/test_apps/rtc_power_modes/README.md": ['esp32', 'esp32s2', 'esp32c3', 'esp32s3', 'esp32c2', 'esp32c6']

If you want to disable some targets, please use the following snippet:

# Please combine this with the original one
#
# Notes:
# - please keep in mind to avoid duplicated folders as yaml keys
# - please use parentheses to group conditions, the "or" and "and" operators could only accept two operands
components/esp_system/test_apps/rtc_power_modes:
  enable:
    - if: IDF_TARGET in ["esp32", "esp32c2", "esp32c3", "esp32c6", "esp32s2", "esp32s3"]
      temporary: true
      reason: <why only enable build jobs for these targets>
check_enable_build failed for app: (cmake) App components/esp_system/test_apps/rtc_power_modes, target esp32, sdkconfig (default), build in /Users/charlesweinberger/Desktop/esp-idf-v5/esp-idf/components/esp_system/test_apps/rtc_power_modes/build
--------------------------------------------------------------------------------
components/console/test_apps/console:
- enable build targets according to the manifest file: ['esp32', 'esp32s2', 'esp32c3', 'esp32s3', 'esp32c2', 'esp32c6']
- enable build targets according to the old Supported Targets table under readme "components/console/test_apps/console/README.md": ['esp32', 'esp32s2', 'esp32c3', 'esp32s3', 'esp32c2', 'esp32c6']

If you want to disable some targets, please use the following snippet:

# Please combine this with the original one
#
# Notes:
# - please keep in mind to avoid duplicated folders as yaml keys
# - please use parentheses to group conditions, the "or" and "and" operators could only accept two operands
components/console/test_apps/console:
  enable:
    - if: IDF_TARGET in ["esp32", "esp32c2", "esp32c3", "esp32c6", "esp32s2", "esp32s3"]
      temporary: true
      reason: <why only enable build jobs for these targets>
check_enable_build failed for app: (cmake) App components/console/test_apps/console, target esp32, sdkconfig (default), build in /Users/charlesweinberger/Desktop/esp-idf-v5/esp-idf/components/console/test_apps/console/build
--------------------------------------------------------------------------------
components/esp_system/test_apps/rtc_8md256:
- enable build targets according to the manifest file: ['esp32', 'esp32s2', 'esp32c3', 'esp32s3', 'esp32c2', 'esp32c6']
- enable build targets according to the old Supported Targets table under readme "components/esp_system/test_apps/rtc_8md256/README.md": ['esp32', 'esp32c2', 'esp32c3', 'esp32s2', 'esp32s3']

If you want to disable some targets, please use the following snippet:

# Please combine this with the original one
#
# Notes:
# - please keep in mind to avoid duplicated folders as yaml keys
# - please use parentheses to group conditions, the "or" and "and" operators could only accept two operands
components/esp_system/test_apps/rtc_8md256:
  enable:
    - if: IDF_TARGET in ["esp32", "esp32c2", "esp32c3", "esp32s2", "esp32s3"]
      temporary: true
      reason: <why only enable build jobs for these targets>
check_enable_build failed for app: (cmake) App components/esp_system/test_apps/rtc_8md256, target esp32, sdkconfig (default), build in /Users/charlesweinberger/Desktop/esp-idf-v5/esp-idf/components/esp_system/test_apps/rtc_8md256/build
--------------------------------------------------------------------------------
Modified README file: components/esp_system/test_apps/rtc_8md256/README.md
--------------------------------------------------------------------------------
components/esp_system/test_apps/rtc_power_modes:
- enable build targets according to the manifest file: ['esp32', 'esp32s2', 'esp32c3', 'esp32s3', 'esp32c2', 'esp32c6']
- enable build targets according to the old Supported Targets table under readme "components/esp_system/test_apps/rtc_power_modes/README.md": ['esp32s3']

If you want to disable some targets, please use the following snippet:

# Please combine this with the original one
#
# Notes:
# - please keep in mind to avoid duplicated folders as yaml keys
# - please use parentheses to group conditions, the "or" and "and" operators could only accept two operands
components/esp_system/test_apps/rtc_power_modes:
  enable:
    - if: IDF_TARGET in ["esp32s3"]
      temporary: true
      reason: <why only enable build jobs for these targets>
check_enable_build failed for app: (cmake) App components/esp_system/test_apps/rtc_power_modes, target esp32, sdkconfig (default), build in /Users/charlesweinberger/Desktop/esp-idf-v5/esp-idf/components/esp_system/test_apps/rtc_power_modes/build
--------------------------------------------------------------------------------
Modified README file: components/esp_system/test_apps/rtc_power_modes/README.md
--------------------------------------------------------------------------------

sort .build-test-rules.yml files.............................................................................(no files to check)Skipped
File Contents Sorter.........................................................................................(no files to check)Skipped
Check copyright notices..........................................................................................................Passed
[WARNING] Stashed changes conflicted with hook auto-fixes... Rolling back fixes...
[INFO] Restored changes from /Users/charlesweinberger/.cache/pre-commit/patch1667903318-74132.
charlesweinberger@Chips-Macbook-Pro-2 esp-idf % git status
On branch user/chip/more-secondary-console-choices-try-2
Your branch is ahead of 'origin/user/chip/more-secondary-console-choices-try-2' by 35 commits.
  (use "git push" to publish your local commits)

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   components/app_trace/Kconfig
	modified:   components/bootloader_support/src/bootloader_console.c
	modified:   components/bootloader_support/src/bootloader_console_loader.c
	modified:   components/console/esp_console.h
	modified:   components/console/esp_console_repl.c
	modified:   components/esp_gdbstub/esp32c3/gdbstub_esp32c3.c
	modified:   components/esp_gdbstub/esp32c6/gdbstub_esp32c6.c
	modified:   components/esp_gdbstub/esp32h2/gdbstub_esp32h2.c
	modified:   components/esp_pm/pm_impl.c
	modified:   components/esp_system/Kconfig
	modified:   components/esp_system/include/esp_private/usb_console.h
	modified:   components/esp_system/panic.c
	modified:   components/esp_system/port/cpu_start.c
	modified:   components/esp_system/port/soc/esp32s2/CMakeLists.txt
	modified:   components/esp_system/port/soc/esp32s2/usb_console.c
	modified:   components/esp_system/port/soc/esp32s3/CMakeLists.txt
	modified:   components/esp_system/port/soc/esp32s3/usb_console.c
	modified:   components/esp_system/startup.c
	modified:   components/vfs/CMakeLists.txt
	modified:   components/vfs/vfs_console.c
	modified:   examples/system/console/advanced/main/console_example_main.c
```

</details>

chipweinberger commented:
I just cloned a fresh esp-idf and am still failing pre-commit locally, even on master branch.

```
Validate executable-list.txt.....................................................................................................Failed
- hook id: check-executable-list
- exit code: 1

Validate Codeowner File..........................................................................................................Failed
- hook id: check-codeowners
- exit code: 1

Check if all apps readme files match given .build-test-rules.yml files. Modify the supported target tables.......................Failed
- hook id: check-all-apps-readmes
- exit code: 1
- files were modified by this hook
```

igrr commented:
sha=c55ff9f2c824555c21be9d338066b63d0dcd6366

igrr commented:
Thank you for the contribution @chipweinberger! I will take it from here and add CI checks for this new feature. If there are any significant issue identified in the internal review, I will get back to you. Otherwise you should get a notification when the PR is merged.

chipweinberger commented:
thanks for the review! ESP-IDF is very high quality code.

igrr commented:
Status update: this got stuck a bit because we needed to add new test environments where both UART and USB are connected to the host, to verify that console output is indeed coming to both destinations.

chipweinberger commented:
Bump. Hoping the testing environment gets setup soon.

chipweinberger commented:
We used to have a `Secondary Usb Serial JTAG output` without having a test environment. 

Maybe skip the new environment if it is difficult.

igrr commented:
The test environment has now been set up, what remains is to write the tests.

chipweinberger commented:
any hope of getting into v5.1? It's been awhile

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/espressif/esp-idf?pullRequest=10132) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/espressif/esp-idf?pullRequest=10132) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/espressif/esp-idf?pullRequest=10132) it.</sub>

chipweinberger commented:
@zikalino , lmk if you need anything from me for this. if you have any questions.

chipweinberger commented:
@igrr, hoping we can clear out some of these PRs.

This PR is useful for 2 reasons:
- some people use the USB Serial/JTAG as their primary console, and therefore should be able also mirror output to UART
- the internal kconfigs for `ESP_CONSOLE_IS_USB_CDC_ENABLED`, etc, which check if either primary or secondary are enabled is a good house-keeping change. simplifies code.

chipweinberger commented:
@igrr will we ever be able to get this in hah =)

It is more of a simplifying change than anything.

chipweinberger commented:
rebased again

chipweinberger commented:
to explain the benefit of this PR to the best of my memory

Console & logging are very limiting.  Output must be chosen at compile time instead of runtime 

and esp-idf only supports uart primary & usb serial jtag secondary. that's very limiting 

this PR makes all logging outputs first class citizens

igrr commented:
@chipweinberger Thanks for patiently keeping this PR alive! I have now added an app to test some of the secondary output options. It should be closer to being merged now.

igrr commented:
@chipweinberger Thanks for fixing that typo. I ran into two more issues while testing this:
- A build error when CDC is selected as secondary output:
   ```
   libesp_system.a(panic.c.obj): in function `panic_print_char_usb_cdc':
       components/esp_system/panic.c:220: undefined reference to `esp_usb_console_write_buf'
   ```
- Bootloader size has increased after these changes, overflowing the 0x8000 max size for a couple of test apps on the ESP32. This is a bit annoying, since the feature doesn't bring much benefit for the ESP32 (it has neither USB CDC nor USB_SERIAL_JTAG), so the fact that the code size has increased is harder to justify. This is not a hard blocker, though, we can probably find a way around this.

chipweinberger commented:
@igrr thanks for testing

> A build error when CDC is selected as secondary output:
> `libesp_system.a(panic.c.obj): in function `panic_print_char_usb_cdc':`
>     `components/esp_system/panic.c:220: undefined reference to `esp_usb_console_write_buf'`

It seems like CDC has changed a lot since this commit was made. I've pushed a new commit. You can see the diff.

> Bootloader size has increased after these changes, overflowing the 0x8000 max size for a couple of test apps on the ESP32. This is a bit annoying, since the feature doesn't bring much benefit for the ESP32 (it has neither USB CDC nor USB_SERIAL_JTAG), so the fact that the code size has increased is harder to justify. This is not a hard blocker, though, we can probably find a way around this.

Even if you leave secondary disabled? I'm not sure why that would be.

ammaree commented:
Can I add +99 here please.
The current kconfig options does not allow UART to be combined with CDC....

igrr commented:
Sure! We haven't forgotten about this. In fact, we [just merged](https://github.com/espressif/esp-idf/commit/a7d1da94b9a70e9dfb8b48ebba0fdd0da5b2658b) a refactoring which should help us simplify how the console devices are managed, without having to add a whole bunch of ifdefs around the codebase. We will definitely revisit this topic once a few other vfs-related refactorings are done. Sorry to keep you all waiting!

mickeyl commented:
This looks great, I'd really like this to land in preparation of better console support. Good work, @chipweinberger!

sidwarkd commented:
@igrr @chipweinberger I know this is over a year later but just wanted to follow up on this to see what options exist currently. I have an S3-based product that has a requirement that the factory can run some REPL tests (input and output) over the onboard USB Serial/JTAG Controller but, for the DUT rigs, need to get just logging output on UART0 for capture and monitoring.

As far as my research has shown today this is still not possible. Is that correct or am I missing something? If it is possible, could you point me in the right direction?

Thanks so much for your continued work on this and for diligently keeping the PR alive Chip.

chipweinberger commented:
what you need to do:

1. uart primary
2. usj secondary
3. check if usj is connected at boot time and if so open the console using usj

that was my workaround solution

sidwarkd commented:
@chipweinberger I'm sorry for being slow to understand here and thank you so much for the help and patience. 

As far as I can tell, that PR was closed and the IDF component `esp_linenoise` is what I would want to use to accomplish my goal. I can see that the config allows me to set the input and output file descriptor. Seems like this should all work but it's not clear from looking through those how I would get the JTAG serial file descriptors. Feels like I'm really close. A reference implementation would be super helpful.

sidwarkd commented:
@chipweinberger @SoucheSouche Try as I might I cannot get this to work in IDF 5.4. I can see the output mirrored on both the USB Serial and UART ports so that is working as I need. :heavy_check_mark:  

I then try to use the `esp_linenoise` managed component to create a console that only works on the USB Serial side (the UART0 side to ESP-Prog is only for log collection). I successfully open the /dev/usbserjtag file and use that file descriptor in the `esp_linenoise_config_t`. I have tried a lot of combinations of only assigning the `in_fd`, setting `out_fd` to `NULL` and other approaches. The application runs and mirrors the output but the second I type a character in the console of USB Serial the program crashes.

@SoucheSouche if you could provide a working example of this setup it would be extremely helpful.

sidwarkd commented:
@SoucheSouche I was finally able to figure it out after going very deep on file descriptors and VFS. I will put together a reference project and share a link here for others. I also will log an issue on the extra components repo as I found a few issues in the esp_linenoise component.

I have mirrored log output on UART0 and USB Serial and a console working on just USB Serial. :heavy_check_mark:
